File: mimo/clipping.m
function [x_clipped,sigma]=clipping(x,CR,sigma)
% CR: Clipping Ratio, sigma: sqrt(variance of x)
if nargin<3
    x_mean=mean(x); x_dev=x-x_mean;
    sigma=sqrt(x_dev*x_dev'/length(x));
end
x_clipped = x; CL = CR*sigma; % Clipping level
ind = find(abs(x)>CL); % Indices to clip
x_clipped(ind)=x(ind)./abs(x(ind))*CL; % Eq.(7.18b) limitation to CL

File: mimo/clipping.py
import numpy as np

def clipping(x, CR, sigma=None):
    """
    Clip signal based on clipping ratio

    Parameters:
    -----------
    x : array_like
        Input signal
    CR : float
        Clipping Ratio
    sigma : float, optional
        Square root of variance of x. If not provided, it will be calculated.

    Returns:
    --------
    x_clipped : ndarray
        Clipped signal
    sigma : float
        Square root of variance (standard deviation)
    """
    x = np.asarray(x)

    if sigma is None:
        x_mean = np.mean(x)
        x_dev = x - x_mean
        sigma = np.sqrt(np.dot(x_dev, np.conj(x_dev)) / len(x))

    x_clipped = x.copy()
    CL = CR * sigma  # Clipping level

    # Find indices where absolute value exceeds clipping level
    ind = np.abs(x) > CL

    # Clip values: normalize and multiply by clipping level
    x_clipped[ind] = x[ind] / np.abs(x[ind]) * CL

    return x_clipped, sigma